### PR TITLE
Run3-hcx360 Make ZDC geometry varying with ERAs 

### PR DESCRIPTION
**Pull request by bsunanda**
#### PR description:

Make ZDC geometry vary with ERAs. It started with an experimental one and matured to become a concrete one.

#### PR validation:

It fails in tests. Made to share it with Yana

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

This works and may be backported to 14_1_X

### CODE REVIEW

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Review by ianna — COMMENTED** [review]
@bsunanda - I think, you need a new GT where a new ZDC payload has IOV from now **and** produce new HCAL payloads (without ZDC and with IOV upto now) then append the new ZDC payload with IOV from now. Hope this helps.

**Comment by antoniovilela** [comment]
ping (to make bot change milestone)

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by civanch** [comment]
@bsunanda , is there any sense to test it now?

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
type bugfix

**Comment by bsunanda** [comment]
@perrotta @civanch @srimanob Please approve this if the test passes

**Comment by perrotta** [comment]
> @perrotta @civanch @srimanob Please approve this if the test passes

Well, tests do no pass yet.
When they will, please remove "(experimental PR not yet working fully)" from the PR title

File: Geometry/ForwardGeometry/src/ZdcHardcodeGeometryLoader.cc [hunk 1]
```diff
@@ -7,13 +7,12 @@
 
 typedef CaloCellGeometry::CCGFloat CCGFloat;
 
+/*
```
Comment (perrotta) [inline Geometry/ForwardGeometry/src/ZdcHardcodeGeometryLoader.cc:10]: Can these commented out parts be just removed, to keep the code clean?

File: Geometry/ForwardGeometry/interface/ZdcHardcodeGeometryLoader.h [hunk 2]
```diff
@@ -30,7 +30,7 @@ class ZdcHardcodeGeometryLoader {
   void fill(HcalZDCDetId::Section section, CaloSubdetectorGeometry* cg);
   void makeCell(const HcalZDCDetId& detId, CaloSubdetectorGeometry* geom) const;
 
-  ZdcTopology* theTopology;
+  //  ZdcTopology* theTopology;
```
Comment (perrotta) [inline Geometry/ForwardGeometry/interface/ZdcHardcodeGeometryLoader.h:33]: please remove

File: Geometry/ForwardGeometry/interface/ZdcHardcodeGeometryLoader.h [hunk 3]
```diff
@@ -17,11 +17,11 @@ class ZdcHardcodeGeometryLoader {
 public:
   typedef CaloSubdetectorGeometry* ReturnType;
 
-  ZdcHardcodeGeometryLoader();
+  //  ZdcHardcodeGeometryLoader();
```
Comment (perrotta) [inline Geometry/ForwardGeometry/interface/ZdcHardcodeGeometryLoader.h:20]: If not needed, please remove

**Review by perrotta — COMMENTED** [review]
Just a few minor comments from a first glance, but since you have to fix already they could be already integrated

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: Geometry/ForwardGeometry/interface/ZdcHardcodeGeometryLoader.h (hunk 3 above)
Comment (bsunanda) [reply Geometry/ForwardGeometry/interface/ZdcHardcodeGeometryLoader.h]: removed

File: Geometry/ForwardGeometry/interface/ZdcHardcodeGeometryLoader.h (hunk 2 above)
Comment (bsunanda) [reply Geometry/ForwardGeometry/interface/ZdcHardcodeGeometryLoader.h:32]: done

File: Geometry/ForwardGeometry/src/ZdcHardcodeGeometryLoader.cc (hunk 1 above)
Comment (bsunanda) [reply Geometry/ForwardGeometry/src/ZdcHardcodeGeometryLoader.cc]: done

File: Geometry/CaloGeometry/src/CaloGenericDetId.cc [hunk 4]
```diff
@@ -44,6 +44,9 @@ uint32_t CaloGenericDetId::sizeForDenseIndexing() const {
   } else if (isCaloTower()) {
     edm::LogError("CaloGenericDetIdError") << "No support for CaloTower in CaloGenericDetId";
     throw cms::Exception("No support");
+    //  } else if (isZDC()) {
```
[the remark this answers is not in the document]
Comment (bsunanda) [reply Geometry/CaloGeometry/src/CaloGenericDetId.cc]: comments removed and they are now used in the code. Further cleanup will be done in the future

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@cmsbuild Please test

File: HLTrigger/Configuration/test/OnLine_HLT_GRun.py [hunk 5]
```diff
@@ -4502,6 +4502,10 @@
 process.ZdcGeometryFromDBEP = cms.ESProducer( "ZdcGeometryFromDBEP",
   applyAlignment = cms.bool( False )
 )
+process.ZdcTopologyEP = cms.ESProducer( "ZdcTopologyEP",
```
Comment (mmusich) [inline HLTrigger/Configuration/test/OnLine_HLT_GRun.py:4505]: please do not change files in `HLTrigger/Configuration` directly, as we ultimately need to update also the configuration in confDB. Please provide the needed `ESProducer` via customization function, see  `HLTrigger/Configuration/python/customizeHLTforCMSSW.py` for examples.

**Comment by mmusich** [comment]
-hlt
  * see https://github.com/cms-sw/cmssw/pull/45511/files#r1760508157

**Comment by bsunanda** [comment]
@mmusich This will affect all future offline processes and I do not know how the internal things are done. Could you let me know how to add the customiseForOffline which has to add the following lines:
process.ZdcTopologyEP = cms.ESProducer( "ZdcTopologyEP",
  Exclude = cms.untracked.string( "" ),
  appendToDataLabel = cms.string( "" )
)
in the scripts. I shall remove the lines from those 4 python scripts and add the supplied lines in HLTrigger/Configuration/python/customizeHLTforCMSSW.py

**Comment by mmusich** [comment]
@bsunanda 

>Could you let me know how to add the customiseForOffline which has to add the following lines:

please take a look at how it was done e.g. [here](https://github.com/cms-sw/cmssw/pull/44576/files#diff-8fc23ee531566af47b6eff1b354bb3db1c4e8f3925413edc3632e417d168b75fR264).

>  I shall remove the lines from those 4 python scripts and add the supplied lines in HLTrigger/Configuration/python/customizeHLTforCMSSW.py

Yes, please

**Comment by bsunanda** [comment]
@cmsbuild Please test

**Comment by bsunanda** [comment]
@Dr15Jones, @Martin-Grunewald, @aloeliger, @bsunanda, @civanch, @consuegs, @epalencia, @francescobrivio, @makortel, @mdhildreth, @mmusich, @perrotta, @saumyaphor4252, @srimanob, The failure are for DAS errors. Can this PR be approved?

**Comment by mmusich** [comment]
@bsunanda 

>  (experimental PR not yet working fully) 

this for sure has to go from the PR title if you want this approved.

**Comment by mmusich** [comment]
> this for sure has to go from the PR title if you want this approved.

also the PR description suffers from the same shortcoming. Please provide an updated version.

**Comment by civanch** [comment]
+1

@bsunanda , please, update description and title.

**Comment by bsunanda** [comment]
@Dr15Jones, @Martin-Grunewald, @aloeliger, @bsunanda, @civanch, @consuegs, @epalencia, @francescobrivio, @makortel, @mdhildreth, @mmusich, @perrotta, @saumyaphor4252, @srimanob, Since the PR has passed and it is urgent for porting back to 14_1_X, could you please sign this? Also, the header is appropriately modified to reflect the reality

**Comment by mmusich** [comment]
+hlt

**Comment by bsunanda** [comment]
@aloeliger, @consuegs, @epalencia, @francescobrivio, @perrotta, @saumyaphor4252, @srimanob, @subirsarkar, Please approve this PR

**Comment by mandrenguyen** [comment]
urgent
Needed for heavy-ion data taking

File: Geometry/ForwardGeometry/plugins/ZdcTopologyEP.cc [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+/**\class ZdcTopologyEP ZdcTopologyEP.h tmp/ZdcTopologyEP/interface/ZdcTopologyEP.h
+
+ Description: <one line class summary>
+
+ Implementation:
+     <Notes on implementation>
+*/
+//
+// Original Author:  Jeremiah Mans
+//         Created:  Mon Oct  3 11:35:27 CDT 2005
```
Comment (perrotta) [inline Geometry/ForwardGeometry/plugins/ZdcTopologyEP.cc:15]: Probably this is a bit outdated here

File: Geometry/ForwardGeometry/plugins/ZdcTopologyEP.cc [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+//
+
+#include "Geometry/ForwardGeometry/plugins/ZdcTopologyEP.h"
+#include "Geometry/HcalCommonData/interface/HcalDDDRecConstants.h"
+#include "FWCore/Utilities/interface/Exception.h"
+#include "FWCore/MessageLogger/interface/MessageLogger.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+
+#define EDM_ML_DEBUG
```
Comment (perrotta) [inline Geometry/ForwardGeometry/plugins/ZdcTopologyEP.cc:26]: Should this remain `define`'d in production mode?

File: Geometry/ForwardGeometry/src/ZdcGeometry.cc [hunk 8]
```diff
@@ -17,21 +17,25 @@ typedef CaloCellGeometry::Tr3D Tr3D;
 
 typedef CaloSubdetectorGeometry::CCGFloat CCGFloat;
 
-//#define EDM_ML_DEBUG
+#define EDM_ML_DEBUG
```
Comment (perrotta) [inline Geometry/ForwardGeometry/src/ZdcGeometry.cc:20]: Again, should this remain `define`'d?

File: CondFormats/HcalObjects/src/HcalCondObjectContainerBase.cc [hunk 9]
```diff
@@ -9,16 +10,24 @@
...
+                                  : HcalZDCDetId::kSizeForDenseIndexingRun1);
+  }
 }
 
 void HcalCondObjectContainerBase::setTopo(const HcalTopology* topo) {
   if ((topo) && (packedIndexVersion_ != 0) && (!topo->denseIdConsistent(packedIndexVersion_)))
     edm::LogError("HCAL") << std::string("Inconsistent dense packing between current topology (") << topo->topoVersion()
                           << ") and calibration object (" << packedIndexVersion_ << ")";
   topo_ = topo;
+  packedIndexVersion_ = topo_->topoVersion();
```
Comment (srimanob) [inline CondFormats/HcalObjects/src/HcalCondObjectContainerBase.cc:28]: >> Analyzing  src/CondFormats/HcalObjects/src/HcalLongRecoParam.cc 
src/CondFormats/HcalObjects/src/HcalCondObjectContainerBase.cc:28:25: warning: Called C++ object pointer is null [core.CallAndMessage]
   28 |   packedIndexVersion_ = topo_->topoVersion();
      |                         ^~~~~~~~~~~~~~~~~~~~
1 warning generated.

**Review by perrotta — COMMENTED** [review]
Even if not in the AlCaDB domain, before giving the green light I would get assured that the unused `m_restrictions` parameter in Geometry/ForwardGeometry/plugins/ZdcTopologyEP.cc was not intended to have some possible usage instead.
All other comments listed here are mostly intended to clean up the code (but perhaps you could at least take care of un-`define` some debug lines).

**Comment by bsunanda** [comment]
Please sign this one and let it merge. I have to define 2024 geometry once it is merged. I can take care of all these comments in that new PR.

**Comment by perrotta** [comment]
As further comments:
- There are several differences in the HCal outputs (and related ones) that are not so obvious to me that they are intended: could you please comment on them?
- Moreover, I notice that there are [less warnings](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-15-2300+6e4b6e/64558/validateJR/logRootQA-events.log) in outputs. They come from `RecoEgamma/EgammaElectronAlgos/src/ElectronSeedGenerator.cc`, and maybe they just witness something that was (inadvertently?) fixed by this PR: could you also please comment on them?

**Comment by perrotta** [comment]
> Please sign this one and let it merge. I have to define 2024 geometry once it is merged. I can take care of all these comments in that new PR.

@sunanda as I wrote: please answer about the unused `m_restrictions` parameter in Geometry/ForwardGeometry/plugins/ZdcTopologyEP.cc , and then I am mostly fine with the code.

Well, please also comment about the differences in outputs: are they expected?

File: Geometry/ForwardGeometry/plugins/ZdcTopologyEP.cc [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+ZdcTopologyEP::ZdcTopologyEP(const edm::ParameterSet& conf)
+    : m_hdcToken{setWhatProduced(this, &ZdcTopologyEP::produce).consumes<HcalDDDRecConstants>(edm::ESInputTag{})},
+      m_restrictions(conf.getUntrackedParameter<std::string>("Exclude")) {
```
[the remark this answers is not in the document]
Comment (bsunanda) [reply Geometry/ForwardGeometry/plugins/ZdcTopologyEP.cc:30]: I shall take care of this during the cleanup at the next PR. I need this merging to take care of the 2024 scenario which is urgently needed.

**Comment by mmusich** [comment]
> You potentially [added 48 lines](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-15-2300+6e4b6e/64558/validateJR/logRootQA-events.log) to the logs

> Moreover, I notice that there are [less warnings](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_14_2_X_2024-09-15-2300+6e4b6e/64558/validateJR/logRootQA-events.log) in outputs. 

@perrotta something is not consistent here.

> They come from RecoEgamma/EgammaElectronAlgos/src/ElectronSeedGenerator.cc, and maybe they just witness something that was (inadvertently?) fixed by this PR: could you also please comment on them?

they look the same amount to me, just with different values. By the way this is linked to the open issue https://github.com/cms-sw/cmssw/issues/45658

**Comment by bsunanda** [comment]
@perrotta I shall remove m_restricions which was there by mistake (I believe copied from heal where it was needed. I have to make a new PR for the 2024 geometry definition and I shall take care of all these comments there.

**Comment by abdoulline** [comment]
Just would like to note that exclusion of the legacy ZDC from Digitization globally reshuffles DIGI rndm...

**Comment by perrotta** [comment]
> @perrotta I shall remove m_restricions which was there by mistake (I believe copied from heal where it was needed. I have to make a new PR for the 2024 geometry definition and I shall take care of all these comments there.

Ok, thank you: this is the answer I needed

**Comment by perrotta** [comment]
> Just would like to note that exclusion of the legacy ZDC from Digitization globally reshuffles DIGI rndm...

Ok, thank you: this is the answer I needed

**Comment by perrotta** [comment]
+1
- The doubts that I had in the [review](https://github.com/cms-sw/cmssw/pull/45511#pullrequestreview-2308665669) were answered in https://github.com/cms-sw/cmssw/pull/45511#issuecomment-2354915490, https://github.com/cms-sw/cmssw/pull/45511#issuecomment-2354918147, https://github.com/cms-sw/cmssw/pull/45511#issuecomment-2354919195
- It is understood that some serious clean up is needed once this code is merged (better if in time to get also backported to CMSSW_14_1_X)

**Comment by aloeliger** [comment]
+l1

**Comment by bsunanda** [comment]
@srimanob @subirsarkar Please approve this PR

**Comment by srimanob** [comment]
+Upgrade

No issue in related upgrade part. Many clean ups to follow, and also null pointer mentioned in https://github.com/cms-sw/cmssw/pull/45511#discussion_r1762604052

**Comment by mandrenguyen** [comment]
+1

**Comment by perrotta** [comment]
> No issue in related upgrade part. Many clean ups to follow, and also null pointer mentioned in [#45511 (comment)](https://github.com/cms-sw/cmssw/pull/45511#discussion_r1762604052)

@srimanob a possible suggestion for that potentially null pointer was provided in https://github.com/cms-sw/cmssw/pull/45511#discussion_r1762625206

**Pull request merged**